### PR TITLE
Shell implementation for Roll operation.

### PR DESCRIPTION
**Pull request by popovaan**
### Details:
 - *nGraph shell for the operation Roll.*

### Tickets:
 - *49386*

### CODE REVIEW

File: ngraph/core/include/ngraph/op/roll.hpp [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+                /// \param      axes         Node producing the 0D or 1D tensor which specifies axes
+                /// along which elements are shifted.
+                ///
+                Roll(const Output<Node>& data, const Output<Node>& shift, const Output<Node>& axes);
+
+                void validate_and_infer_types() override;
+
+                bool visit_attributes(AttributeVisitor& visitor) override;
+
+                virtual std::shared_ptr<Node>
```
Comment (ilyachur) [inline ngraph/core/include/ngraph/op/roll.hpp:51]: virtual is redundant.

**Review by achetver — APPROVED** [review]
Looks good to me

**Review by ilyachur — APPROVED** [review]

**Review by sadolini — APPROVED** [review]

File: ngraph/core/src/op/roll.cpp [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+    const auto& shift_pshape = get_input_partial_shape(1);
+    const auto& axes_pshape = get_input_partial_shape(2);
+    const auto& shift_rank = shift_pshape.rank().get_length();
+    const auto& axes_rank = axes_pshape.rank().get_length();
+
+    NODE_VALIDATION_CHECK(this, shift_rank <= 1, "Shift must be a scalar or 1D tensor.");
+
+    NODE_VALIDATION_CHECK(this, axes_rank <= 1, "Axes must be a scalar or 1D tensor.");
+
+    if (!(shift_pshape.is_static() &&
```
Comment (pavel-esir) [inline ngraph/core/src/op/roll.cpp:56]: This check looks too complicated to me. I understand that it goes for https://github.com/openvinotoolkit/openvino/pull/4469/files#diff-2b0119dc5b43a7eb3a925bfddabd9f0fb59890203d7fa0fbe1b0375c706bb3f5R58

AFAIU this if statement means:
>No need to check `shift` shape consistency with `axes` if `shift` is a scalar or 1D tensor with a single value . And always check consistency if `shift` partial_shape is dynamic.

Maybe adding comment will be appropriate but i'm not insisting on that. 

What is more important is that in the code scalar `shift` and `shift` with shape [1] behave the same way but spec says:
>If `shift` is a 1D tensor, `axes` must be a 1D tensor of the same size

If I understood correctly either the code should be corrected or the specification.

File: ngraph/test/type_prop/roll.cpp [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+TEST(type_prop, roll_dynamic_shape)
+{
+    auto arg = make_shared<op::Parameter>(element::f32,
+                                          PartialShape{Dimension::dynamic(), Dimension::dynamic()});
+    auto axes = make_shared<op::Parameter>(element::i64, PartialShape{Dimension::dynamic()});
+    auto shift = make_shared<op::Parameter>(element::i32, PartialShape{Dimension::dynamic()});
+
+    auto r = make_shared<op::v7::Roll>(arg, axes, shift);
```
Comment (pavel-esir) [inline ngraph/test/type_prop/roll.cpp:79]: According to the spec and to `Roll` constructor `shift` go first then `axes`. Here is a type. Please correct.

File: ngraph/test/type_prop/roll.cpp [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+using namespace std;
+using namespace ngraph;
+
+TEST(type_prop, roll_output_shape_type_test)
+{
+    auto arg = make_shared<op::Parameter>(element::f32, Shape{3, 3, 4, 1, 5});
+    auto axes = make_shared<op::Parameter>(element::i64, Shape{2});
+    auto shift = make_shared<op::Parameter>(element::i32, Shape{2});
+
+    auto r = make_shared<op::v7::Roll>(arg, axes, shift);
```
Comment (pavel-esir) [inline ngraph/test/type_prop/roll.cpp:30]: According to the spec and to Roll constructor shift go first then axes. Here is a type. Please correct. 
Same for lines 52, 56 and so on.

**Review by pavel-esir — CHANGES_REQUESTED** [review]

File: ngraph/core/src/op/roll.cpp (hunk 2 above)
Comment (popovaan) [reply ngraph/core/src/op/roll.cpp]: Added comment and removed 1d tensor with a single value check.

File: ngraph/core/src/op/roll.cpp [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+    constructor_validate_and_infer_types();
+}
+
+void op::v7::Roll::validate_and_infer_types()
+{
+    NGRAPH_OP_SCOPE(v7_Roll_validate_and_infer_types);
+
+    const auto& shift_et = get_input_element_type(1);
+    NODE_VALIDATION_CHECK(this,
+                          shift_et.is_dynamic() || shift_et.is_integral_number(),
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/roll.cpp:39]: According to the specification `shift` and `axes` inputs can be of type `int32` or `int64` only. Please, change these checks.

**Review by pavel-esir — APPROVED** [review]

File: ngraph/core/src/op/roll.cpp [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+    NODE_VALIDATION_CHECK(this,
+                          axes_et.is_dynamic() || axes_et.is_integral_number(),
+                          "Axes must have an integral number element type.");
+
+    const auto& data_pshape = get_input_partial_shape(0);
+    const auto& shift_pshape = get_input_partial_shape(1);
+    const auto& axes_pshape = get_input_partial_shape(2);
+    const auto& data_rank = data_pshape.rank().get_length();
+    const auto& shift_rank = shift_pshape.rank().get_length();
+    const auto& axes_rank = axes_pshape.rank().get_length();
```
Comment (lazarevevgeny) [inline ngraph/core/src/op/roll.cpp:52]: Rank can be dynamic. So the `get_length()` will fail. Please, add a check that the rank is not dynamic before getting its length

Comment (lazarevevgeny) [reply ngraph/core/src/op/roll.cpp:52]: Please, add unit tests for this case.

File: ngraph/test/type_prop/roll.cpp [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+#include "gtest/gtest.h"
+#include "ngraph/ngraph.hpp"
+#include "util/type_prop.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+TEST(type_prop, roll_output_shape_type_test)
+{
+    auto arg = make_shared<op::Parameter>(element::f32, Shape{3, 3, 4, 1, 5});
```
Comment (lazarevevgeny) [inline ngraph/test/type_prop/roll.cpp:26]: Do not use `op::OpName` approach. Instead use `opsetX::OpName` approach. In this case use `opset7` everywhere. Please, fix all test cases.

File: ngraph/test/type_prop/roll.cpp [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+            error.what(),
+            std::string("If shift is a 1D vector, axes must be a 1D tensor of the same size."));
+    }
+    catch (...)
+    {
+        FAIL() << "Check failed for unexpected reason";
+    }
+}
+
+TEST(type_prop, roll_dynamic_shape)
```
Comment (lazarevevgeny) [inline ngraph/test/type_prop/roll.cpp:128]: Add test cases with a mix of dynamic shape of `shift` and `axes`. For example, `shift` it dynamic while `axes` is not, and vice-versa.

**Review by lazarevevgeny — CHANGES_REQUESTED** [review]

**Review by lazarevevgeny — APPROVED** [review]

**Pull request merged**